File: part-c-autoresearch/plugins/autoresearch.mjs
/**
 * Part C custom DeepSeek Harness plugin.
 *
 * The plugin exposes one bounded, model-callable tool. It delegates the
 * numerical benchmark to a local Python process, validates all user inputs,
 * and returns a structured experiment report. No arbitrary command, path, or
 * code is accepted from the model.
 */

import { spawn } from 'node:child_process'
import { fileURLToPath } from 'node:url'
import path from 'node:path'

export const name = 'part-c-autoresearch-ml'
export const inject = ['tools']

const pluginDirectory = path.dirname(fileURLToPath(import.meta.url))
const partDirectory = path.resolve(pluginDirectory, '..')
const benchmarkScript = path.join(partDirectory, 'benchmark', 'autoresearch.py')
const resultsDirectory = path.join(partDirectory, 'results')

function integerInRange(value, fallback, minimum, maximum) {
  if (value === undefined) return fallback
  if (!Number.isInteger(value) || value < minimum || value > maximum) {
    throw new Error(`Value must be an integer between ${minimum} and ${maximum}.`)
  }
  return value
}

function pythonCommand() {
  if (process.env.PYTHON) return process.env.PYTHON
  return process.platform === 'win32' ? 'python' : 'python3'
}

function executeBenchmark({ maxTrials, seed }) {
  return new Promise((resolve, reject) => {
    const child = spawn(
      pythonCommand(),
      [benchmarkScript, '--max-trials', String(maxTrials), '--seed', String(seed), '--output-dir', resultsDirectory, '--json'],
      { cwd: partDirectory, shell: false },
    )

    let stdout = ''
    let stderr = ''
    const timeout = setTimeout(() => {
      child.kill()
      reject(new Error('Autoresearch benchmark exceeded the 30-second safety timeout.'))
    }, 30_000)

    child.stdout.on('data', (chunk) => { stdout += chunk })
    child.stderr.on('data', (chunk) => { stderr += chunk })
    child.on('error', (error) => {
      clearTimeout(timeout)
      reject(new Error(`Could not start Python benchmark: ${error.message}`))
    })
    child.on('close', (code) => {
      clearTimeout(timeout)
      if (code !== 0) {
        reject(new Error(stderr.trim() || `Python benchmark exited with code ${code}.`))
        return
      }
      try {
        resolve(JSON.parse(stdout))
      } catch (error) {
        reject(new Error(`Benchmark returned invalid JSON: ${error.message}`))
      }
    })
  })
}

export function apply(ctx) {
  ctx.tools.register({
    name: 'run_autoresearch_experiment',
    description:
      'Run a bounded ML autoresearch loop, compare validation MSE against a baseline, and return the best reproducible configuration.',
    parameters: {
      type: 'object',
      properties: {
        maxTrials: {
          type: 'integer',
          minimum: 1,
          maximum: 8,
          description: 'Maximum candidate configurations to evaluate. The baseline is always evaluated separately.',
        },
        seed: {
          type: 'integer',
          minimum: 0,
          maximum: 999999,
          description: 'Seed for the deterministic synthetic dataset and split.',
        },
      },
      additionalProperties: false,
    },
    output: {
      schema: {
        type: 'object',
        additionalProperties: true,
        properties: {
          runId: { type: 'string' },
          objective: { type: 'string' },
          baseline: { type: 'object' },
          trials: { type: 'array' },
          best: { type: 'object' },
          improvement: { type: 'number' },
          artifactPath: { type: 'string' },
        },
        required: ['runId', 'objective', 'baseline', 'trials', 'best', 'improvement', 'artifactPath'],
      },
      render: (_args, value) => [
        {
          type: 'text',
          text: [
            `Autoresearch run ${value.runId} completed.`,
            `Baseline validation MSE: ${value.baseline.validationMse.toFixed(4)}`,
            `Best validation MSE: ${value.best.validationMse.toFixed(4)}`,
            `Improvement: ${(value.improvement * 100).toFixed(2)}%`,
            `Artifact: ${value.artifactPath}`,
          ].join('\n'),
        },
      ],
    },
    async execute(args = {}) {
      const maxTrials = integerInRange(args.maxTrials, 5, 1, 8)
      const seed = integerInRange(args.seed, 7, 0, 999999)
      return executeBenchmark({ maxTrials, seed })
    },
  })
}

File: part-c-autoresearch/benchmark/autoresearch.py
"""Bounded autoresearch loop for the Part C DeepSeek Harness plugin."""

from __future__ import annotations

import argparse
import json
import time
import uuid
from pathlib import Path

from train import Config, train_and_evaluate


OBJECTIVE = "minimize_validation_mse"
BASELINE = Config(feature_degree=1, learning_rate=0.03, epochs=20, l2=0.1)
CANDIDATES = [
    Config(feature_degree=1, learning_rate=0.03, epochs=40, l2=0.1),
    Config(feature_degree=2, learning_rate=0.01, epochs=80, l2=0.001),
    Config(feature_degree=2, learning_rate=0.03, epochs=80, l2=0.001),
    Config(feature_degree=2, learning_rate=0.08, epochs=80, l2=0.001),
    Config(feature_degree=2, learning_rate=0.03, epochs=120, l2=0.01),
    Config(feature_degree=2, learning_rate=0.015, epochs=160, l2=0.001),
    Config(feature_degree=2, learning_rate=0.05, epochs=120, l2=0.005),
    Config(feature_degree=1, learning_rate=0.08, epochs=120, l2=0.01),
]


def run_search(seed: int, max_trials: int) -> dict:
    baseline = train_and_evaluate(BASELINE, seed)
    trials = []
    best = baseline
    for index, config in enumerate(CANDIDATES[:max_trials], start=1):
        result = train_and_evaluate(config, seed)
        result["trial"] = index
        result["accepted"] = result["validationMse"] < best["validationMse"]
        if result["accepted"]:
            best = result
        trials.append(result)

    improvement = (baseline["validationMse"] - best["validationMse"]) / baseline["validationMse"]
    run_id = time.strftime("%Y%m%dT%H%M%SZ", time.gmtime()) + f"-{uuid.uuid4().hex[:8]}"
    return {
        "runId": run_id,
        "benchmark": "synthetic-nonlinear-regression",
        "objective": OBJECTIVE,
        "seed": seed,
        "budget": {
            "maxTrials": max_trials,
            "evaluatedTrials": len(trials),
            "candidatePool": len(CANDIDATES),
        },
        "baseline": baseline,
        "trials": trials,
        "best": best,
        "improvement": round(improvement, 8),
        "stoppingReason": "trial_budget_exhausted",
        "safety": {
            "externalNetwork": False,
            "arbitraryCodeExecution": False,
            "maxTrials": 8,
            "fixedDatasetGeneration": True,
        },
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--max-trials", type=int, default=5)
    parser.add_argument("--seed", type=int, default=7)
    parser.add_argument("--output-dir", type=Path, default=Path(__file__).resolve().parents[1] / "results")
    parser.add_argument("--json", action="store_true", help="Print the report as JSON.")
    args = parser.parse_args()
    if not 1 <= args.max_trials <= 8:
        parser.error("--max-trials must be between 1 and 8")
    if not 0 <= args.seed <= 999999:
        parser.error("--seed must be between 0 and 999999")

    report = run_search(seed=args.seed, max_trials=args.max_trials)
    args.output_dir.mkdir(parents=True, exist_ok=True)
    artifact = args.output_dir / f"run-{report['runId']}.json"
    report["artifactPath"] = artifact.as_posix()
    artifact.write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8")

    if args.json:
        print(json.dumps(report))
    else:
        print(f"Run: {report['runId']}")
        print(f"Baseline validation MSE: {report['baseline']['validationMse']:.6f}")
        print(f"Best validation MSE: {report['best']['validationMse']:.6f}")
        print(f"Improvement: {report['improvement'] * 100:.2f}%")
        print(f"Artifact: {report['artifactPath']}")


if __name__ == "__main__":
    main()
